challenges/08-coding-interview-prep/rosetta-code.json

```diff
@@ -12434,6 +12434,1022 @@
+        "<tr><th>Operator</th><th>Precedence</th><th>Associativity</th><th>Operation</th></tr>",
+        "<tr><td>^</td><td>4</td><td>Right</td><td>Exponentiation</td></tr>",
+        "<tr><td>*</td><td>3</td><td>Left</td><td>Multiplication</td></tr>",
+        "<tr><td>/</td><td>3</td><td>Left</td><td>Division</td></tr>",
+        "<tr><td>+</td><td>2</td><td>Left</td><td>Addition</td></tr>",
+        "<tr><td>-</td><td>2</td><td>Left</td><td>Subtraction</td></tr>",
+        "</table>"
+      ],
+      "solutions": [
+        "function shuntingYard (infix) {\n  function Stack() {\n    this.dataStore = [];\n    this.top = 0;\n    this.push = push;\n    this.pop = pop;\n    this.peek = peek;\n    this.length = length;\n  }\n\n  function push(element) {\n    this.dataStore[this.top++] = element;\n  }\n\n  function pop() {\n    return this.dataStore[--this.top];\n  }\n\n  function peek() {\n    return this.dataStore[this.top - 1];\n  }\n\n  function length() {\n    return this.top;\n  }\n\n  infix = infix.replace(/\\s+/g, ''); // remove spaces, so infix[i]!=\" \"\n\n  var s = new Stack();\n  var ops = \"-+/*^\";\n  var precedence = {\n    \"^\": 4,\n    \"*\": 3,\n    \"/\": 3,\n    \"+\": 2,\n    \"-\": 2\n  };\n  var associativity = {\n    \"^\": \"Right\",\n    \"*\": \"Left\",\n    \"/\": \"Left\",\n    \"+\": \"Left\",\n    \"-\": \"Left\"\n  };\n  var token;\n  var postfix = \"\";\n  var o1, o2;\n\n  for (var i = 0; i < infix.length; i++) {\n    token = infix[i];\n    if (token >= \"0\" && token <= \"9\") { // if token is operand (here limited to 0 <= x <= 9)\n      postfix += token + \" \";\n    } else if (ops.indexOf(token) != -1) { // if token is an operator\n      o1 = token;\n      o2 = s.peek();\n      while (ops.indexOf(o2) != -1 && ( // while operator token, o2, on top of the stack\n          // and o1 is left-associative and its precedence is less than or equal to that of o2\n          (associativity[o1] == \"Left\" && (precedence[o1] <= precedence[o2])) ||\n          // the algorithm on wikipedia says: or o1 precedence < o2 precedence, but I think it should be\n          // or o1 is right-associative and its precedence is less than that of o2\n          (associativity[o1] == \"Right\" && (precedence[o1] < precedence[o2]))\n        )) {\n        postfix += o2 + \" \"; // add o2 to output queue\n        s.pop(); // pop o2 of the stack\n        o2 = s.peek(); // next round\n      }\n      s.push(o1); // push o1 onto the stack\n    } else if (token == \"(\") { // if token is left parenthesis\n      s.push(token); // then push it onto the stack\n    } else if (token == \")\") { // if token is right parenthesis\n      while (s.peek() != \"(\") { // until token at top is (\n        postfix += s.pop() + \" \";\n      }\n      s.pop(); // pop (, but not onto the output queue\n    }\n  }\n  postfix += s.dataStore.reverse().join(\" \");\n  return postfix\n}\n"
+      ],
+      "tests": [
```

Comment:
Also, this is a pretty minor detail, but I noticed that many, if not all of the text for each test has single quotes, which looks a bit off on the page. I know it's a lot of work, but could it be changed to the following format?

```
{
          "text": "<code>shuntingYard</code> should be a function.",
          "testString":
            "assert(typeof shuntingYard=='function','<code>shuntingYard</code> should be a function.');"
        },
```

Seems like it's just the text strings that need to be edited.